json_serializable/lib/src/json_serializable_generator.dart

```diff
@@ -129,26 +134,91 @@ class JsonSerializableGenerator
         buffer.writeln('  ${field.type} get ${field.name};');
       }
 
-      var includeIfNull = annotation.read('includeIfNull').boolValue;
+      var classIncludeIfNull = annotation.read('includeIfNull').boolValue;
 
-      buffer.writeln('  Map<String, dynamic> toJson() ');
-      if (fieldsList.every((e) => _includeIfNull(e, includeIfNull))) {
-        // write simple `toJson` method that includes all keys...
-        _writeToJsonSimple(buffer, fields.values);
+      buffer.write('  Map<String, dynamic> toJson() ');
+
+      if (_doNewThing) {
+        buffer.writeln('=> new $helpClassName(this);');
       } else {
-        // At least one field should be excluded if null
-        _writeToJsonWithNullChecks(buffer, fields.values, includeIfNull);
+        if (fieldsList.every((e) => _includeIfNull(e, classIncludeIfNull))) {
+          // write simple `toJson` method that includes all keys...
+          _writeToJsonSimple(buffer, fields.values);
+        } else {
+          // At least one field should be excluded if null
+          _writeToJsonWithNullChecks(buffer, fields.values, classIncludeIfNull);
+        }
       }
 
       // end of the mixin class
-      buffer.write('}');
+      buffer.writeln('}');
+
+      if (_doNewThing) {
+        buffer.writeln();
+        //
+        // Generate the Map helper class
+        //
+        // TODO(kevmoo): write JsonMapWrapper if annotation lib is prefix-imported
+        buffer.writeln('''class $helpClassName extends JsonMapWrapper {
+  final $mixClassName _v;
+  $helpClassName(this._v);
+''');
+
+        if (fieldsList.every((e) => _includeIfNull(e, classIncludeIfNull))) {
+          // write simple `toJson` method that includes all keys...
+          //_writeToJsonSimple(buffer, fields);
+
+          var jsonKeys = fields.values.map(_safeNameAccess).join(', ');
+
+          // TODO(kevmoo): maybe put this in a static field instead?
+          //               const lists have unfortunate overhead
```

Comment:
A static final field would be lazy initialized right? That might help startup performance.